modules/diskq/qdisk.c

```diff
@@ -978,18 +978,17 @@ _save_queue(QDisk *self, GQueue *q, QDiskQueuePosition *q_pos)
 
   serialized = g_string_sized_new(4096);
   sa = serialize_string_archive_new(serialized);
-  while ((msg = g_queue_pop_head(q)))
+  for (gint i = 0; i < g_queue_get_length(q); i+=2)
     {
-
       /* NOTE: we might have some flow-controlled events on qout, when
        * saving them to disk, we ack them, they are restored as
        * non-flow-controlled entries later, but then we've saved them to
        * disk anyway. */
 
-      POINTER_TO_LOG_PATH_OPTIONS(g_queue_pop_head(q), &path_options);
+      msg = g_queue_peek_nth(q, i);
+      POINTER_TO_LOG_PATH_OPTIONS(g_queue_peek_nth(q, i+1), &path_options);
       log_msg_serialize(msg, sa, 0);
-      log_msg_ack(msg, &path_options, AT_PROCESSED);
-      log_msg_unref(msg);
```

Comment:
I think this patch needs to be explained carefully.

According to my previous comments, I think that queues should not always be emptied when the disk-buffer is stopped (corrupt cases). Was this a double-ack/unref issue prior to your changes?

Reply:
Prior to my commits, the messages were freed in `_save_queue()` -> `qdisk_save_state()` -> `_save_queue()`, then there were unnecessary frees in `_free()` -> `_free_queue()`, and no one decreased the memory usage counter.

With my commits the messages are only freed in `_stop()` -> `empty_queue()`, where the memory usage counter is decreased.

I will add a verbose commit message, because it is hard to follow my changes through different commits.

Reply:
```
qdisk: do not consume messages in save_state()

Prior to my this PR, the messages were freed in
_save_queue() -> qdisk_save_state() -> _save_queue(), then there were
unnecessary frees in _free() -> _free_queue(), and no one decreased the
memory usage counter.

With the previous commits I have moved the freeing to _stop(), and the
memory usage is decreased there as well, but the qdisk_save_state()
branch frees it before we try to do it in _stop(), so there are no
messages there to free, and we cannot correctly decrease the memory
usage counter.

This commit changes this, so we no longer free the messages in
qdisk_save_state(), so they are only freed in _stop() -> empty_queue(),
where the memory usage counter is decreased.
```

Reply:
Thanks.